src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,385 @@
+  tprod_coeff R z₁ f + tprod_coeff R z₂ f = tprod_coeff R (z₁ + z₂) f :=
+quotient.sound' $ add_con_gen.rel.of _ _ (eqv.of_add_scalar z₁ z₂ f)
+
+lemma smul_tprod_coeff_aux (z : R) (f : Π i, s i) (i : ι) (r : R) :
+  tprod_coeff R z (update f i (r • f i)) = tprod_coeff R (r * z) f :=
+ quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_smul _ _ _ _
+
+lemma smul_tprod_coeff (z : R) (f : Π i, s i) (i : ι) (r : R')
+  [semimodule R' (s i)] [is_scalar_tower R' R (s i)] :
+  tprod_coeff R z (update f i (r • f i)) = tprod_coeff R (r • z) f :=
+begin
+  have h₁ : r • z = (r • (1 : R)) * z := by simp,
+  have h₂ : r • (f i) = (r • (1 : R)) • f i := by simp,
+  rw [h₁, h₂],
+  exact smul_tprod_coeff_aux z f i _,
+end
+
+/-- Construct an `add_monoid_hom` from `(⨂[R] i, s i)` to some space `F` from a function
+`φ : (R × Π i, s i) → F` with the appropriate properties. -/
+def lift_add_hom (φ : (R × Π i, s i) → F)
+  (C0 : ∀ (r : R) (f : Π i, s i) (i : ι) (hf : f i = 0), φ (r, f) = 0)
+  (C0' : ∀ (f : Π i, s i), φ (0, f) = 0)
+  (C_add : ∀ (r : R) (f : Π i, s i) (i : ι) (m₁ m₂ : s i),
+    φ (r, update f i m₁) + φ (r, update f i m₂) = φ (r, update f i (m₁ + m₂)))
+  (C_add_scalar : ∀ (r r' : R) (f : Π i, s i),
+    φ (r , f) + φ (r', f) = φ (r + r', f))
+  (C_smul : ∀ (r : R) (f : Π i, s i) (i : ι) (r' : R),
+    φ (r, update f i (r' • (f i))) = φ (r' * r, f))
+: (⨂[R] i, s i) →+ F :=
+(add_con_gen (pi_tensor_product.eqv R s)).lift (free_add_monoid.lift φ) $ add_con.add_con_gen_le $
+λ x y hxy, match x, y, hxy with
+| _, _, (eqv.of_zero r' f i hf)        := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C0 r' f i hf]
+| _, _, (eqv.of_zero_scalar f)        := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C0']
+| _, _, (eqv.of_add z f i m₁ m₂)      := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_add]
+| _, _, (eqv.of_add_scalar z₁ z₂ f)      := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_add_scalar]
+| _, _, (eqv.of_smul z f i r')     := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_smul]
+| _, _, (eqv.add_comm x y)         := (add_con.ker_rel _).2 $
+    by simp_rw [add_monoid_hom.map_add, add_comm]
+end
+
+-- Most of the time we want the instance below this one, which is easier for typeclass resolution
+-- to find.
+instance has_scalar' : has_scalar R' (⨂[R] i, s i) :=
+⟨λ r, lift_add_hom (λ f : R × Π i, s i, tprod_coeff R (r • f.1) f.2)
+  (λ r' f i hf, by simp_rw [zero_tprod_coeff' _ f i hf])
+  (λ f, by simp [zero_tprod_coeff])
+  (λ r' f i m₁ m₂, by simp [add_tprod_coeff])
+  (λ r' r'' f, by simp [add_tprod_coeff', mul_add])
+  (λ z f i r', by simp [smul_tprod_coeff])⟩
+
+instance : has_scalar R (⨂[R] i, s i) := pi_tensor_product.has_scalar'
+
+lemma smul_tprod_coeff' (r : R') (z : R) (f : Π i, s i) :
+  r • (tprod_coeff R z f) = tprod_coeff R (r • z) f := rfl
+
+protected theorem smul_zero (r : R') : (r • 0 : ⨂[R] i, s i) = 0 :=
```

Comment:
```suggestion
@[simp] protected theorem smul_zero (r : R') : (r • 0 : ⨂[R] i, s i) = 0 :=
```

Reply:
Is there any point? Can't the global `smul_zero` simp this anyway? This lemma is just an implementation detail needed to build the module structure.

Reply:
I agree with Eric here: this lemma is obsolete by the time the semimodule instance is proven a few lines below. And I just noticed that it's only used once, to prove `smul_zero` in the semimodule instance, so I deleted it.